.dotnet/OpenAI.Microsoft.Generator.CSharp.ClientModel.Plugin/OpenAILibraryVisitor.cs

```diff
@@ -0,0 +1,174 @@
+            }
+            return field;
+        }
+
+        protected override MethodProvider Visit(MethodProvider method)
+        {
+            if (method.Signature.Name != "JsonModelWriteCore")
+            {
+                return method;
+            }
+
+            // If there are no body statements, return the method as is
+            if (method.BodyStatements == null)
+            {
+                return method;
+            }
+
+            // If the body statements are not MethodBodyStatements, return the method as is
+            if (method.BodyStatements is not MethodBodyStatements statements)
+            {
+                return method;
+            }
+
+            var updatedStatements = new List<MethodBodyStatement>();
+
+            foreach (var statement in Flatten(statements!))
+            {
+                if (statement is IfStatement ifStatement)
+                {
+                    var body = ifStatement.Body.ToDisplayString();
+                    if (body.Contains(WritePropertyNameMethodCall))
+                    {
+                        ifStatement.Condition = ifStatement.Condition.As<bool>().And(GetContainsKeyCondition(body));
+                        updatedStatements.Add(ifStatement);
+                        continue;
+                    }
+
+                    if (Flatten(ifStatement.Body).First() is ForeachStatement foreachStatement)
+                    {
+                        foreachStatement.Body.Insert(
+                            0,
+                            new IfStatement(
+                                Static(new ModelSerializationExtensionsDefinition().Type).Invoke("IsSentinelValue", foreachStatement.ItemVariable.Property("Value")))
+                            {
+                                Continue
+                            });
+                    }
+                }
+
+                var displayString = statement.ToDisplayString();
+                if (displayString.Contains(WritePropertyNameMethodCall))
+                {
+                    var ifSt = new IfStatement(GetContainsKeyCondition(displayString)) { statement };
+                    updatedStatements.Add(ifSt);
+                }
+                else
+                {
+                    updatedStatements.Add(statement);
+                }
+            }
+            method.Update(bodyStatements: updatedStatements);
+            return method;
+        }
+
+        private static ScopedApi<bool> GetContainsKeyCondition(string displayString)
+        {
+            var propertyName = displayString.Split('"')[1];
+            return AdditionalBinaryDataExpression
+                .NullConditional()
+                .Invoke("ContainsKey", Literal(propertyName)).NotEqual(True);
+        }
+
+        private IEnumerable<MethodBodyStatement> Flatten(MethodBodyStatement bodyStatement)
+        {
+            if (bodyStatement is MethodBodyStatements statements)
```

Comment:
This looks good for the use case here. I'm wondering what improvements we can _eventually_ make to our statement /+ expression model so this can be done more easily 😄